lib/src/Pos/Network/Block/Retrieval.hs

```diff
@@ -341,16 +341,37 @@ streamProcessBlocks
     -> m ()
 streamProcessBlocks genesisConfig txpConfig diffusion nodeId desired checkpoints = do
     logInfo "streaming start"
-    r <- Diffusion.streamBlocks diffusion nodeId desired checkpoints writeCallback
+    mostDifficultBlock <- atomically $ newTVar Nothing
+    r <- Diffusion.streamBlocks diffusion nodeId desired checkpoints (writeCallback mostDifficultBlock)
     case r of
          Nothing -> do
              logInfo "streaming not supported, reverting to batch mode"
              getProcessBlocks genesisConfig txpConfig diffusion nodeId desired checkpoints
          Just _  -> do
              logInfo "streaming done"
-             return ()
+
+             recHeaderVar <- view (lensOf @RecoveryHeaderTag)
+             exitedRecovery <- atomically $ do
+                 mbMostDifficult <- readTVar mostDifficultBlock
+                 mbRecHeader <- tryReadTMVar recHeaderVar
+                 case (mbMostDifficult, mbRecHeader) of
+                      (Nothing, _)                 -> pure False -- We have not gotten a single block
+                      (Just _, Nothing)            -> pure False -- We where not in recovery?
```

Comment:
RE the `?` in the comment: can we assure ourselves that it true? It makes good sense. I was once more familiar with the recovery mode stuff but now I'd have to go back and check. IIRC the recovery header is set whenever we hear of a header for which we do not know the parent.

Reply:
We should not be streaming blocks if we aren't in recovery mode.